include/openrave/openrave.h

```diff
@@ -1076,7 +1076,15 @@ inline T NormalizeCircularAnglePrivate(T theta, T min, T max)
     }
     return theta;
 }
-
+/// \brief ensures first three elements of vector forms a unit length vector
+/// first checks if input direction vector is close to unit length, and raise if not because it is most likely a bug on application side
+/// \param direction[inout] direction vector to be checked, if far from being unit length, most likely a bug on application side so assert. Otherwise, normalize and return.
+inline void Ensure3DVecUnitLength(Vector& direction)
+{
+    const dReal length2 = direction.lengthsqr3();
+    MATH_ASSERT(length2 > 0.99 && length2 < 1.01); // make sure it is at least close
```

Comment:
@kanbouchou problem with this is that you don't get the stack of which function called it. Doing a MACRO would be better...

Reply:
yeah, I thought about it but decided to go with inline function thinking most env has only one or two ik parameterization types so should be relatively easy to guess the stack.

Sure, I'll make this a macro and update this MR.

Reply:
@rdiankov please review again, I made checker function to a macro (added similar one for quaternion). locally tested that stack (asserted location) is actually the caller of the macro.